### PR TITLE
Issue 80: Updated tests to work on DOM instead of text

### PR DESCRIPTION
**Pull request by SebastianZ**
All tests were recreated to work on the DOM instead of text. This also required to recreate the unit tests and let them run on the document.

Because this is a huge change, it should be reviewed carefully.
I think it makes sense to make a new release first before merging this.

Sebastian

### CODE REVIEW

File: data/doctests.js [hunk 1]
```diff
@@ -22,15 +33,27 @@ var docTests = {
...
-        if (matches[i].match(/^<(?:link|track|param|area|command|col|base|meta|hr|source|img|keygen|br|wbr|input)/)) {
-          matches.splice(i, 1);
-        }
+    check: function checkEmptyElements(rootElement) {
+      var treeWalker = document.createTreeWalker(
+          rootElement,
+          NodeFilter.SHOW_ELEMENT,
+          {
+            acceptNode: (node) => {
+              //alert(node.localName + ", " + node.textContent + ", " + node.localName.match(/^(?:link|track|param|area|command|col|base|meta|hr|source|img|keygen|br|wbr|input)$/i));
```
Comment (Elchi3) [inline data/doctests.js]: I think removing the `alert` and explaining that this excludes self-closing elements is nicer.

File: test/test-doctests.js [hunk 2]
```diff
@@ -1,536 +1,586 @@
...
+
+const url = "https://developer.mozilla.org/en-US/";
+
+function runTests(assert, done, name, desc, url, tests) {
+  tabs.open({
+    url: url,
+    onReady: tab => {
+      var worker = tabs.activeTab.attach({
+        contentScriptFile: [data.url("doctests.js")],
+        contentScript: "var testObj = docTests[self.options.name];" +
```
Comment (Elchi3) [inline test/test-doctests.js]: Could this be another contentScriptFile? I am not really a fan of this string notation.

File: test/test-doctests.js [hunk 3]
```diff
@@ -1,536 +1,586 @@
...
+                       "var rootElement = document.createElement('div');" +
+                       "var tests = JSON.parse(self.options.tests);" +
+                       "tests.forEach(test => {" +
+                       "  rootElement.innerHTML = test.str;" +
```
Comment (Elchi3) [inline test/test-doctests.js]: Is this safe?

File: test/test-doctests.js [hunk 4]
```diff
@@ -1,536 +1,586 @@
...
+        contentScriptOptions: {"name": name, "tests": JSON.stringify(tests)}
+      });
+
+      var resultCount = 0;
+
+      worker.port.on("testResult", function(testObj) {
+        var matches = testObj.errors;
+        var expected = testObj.expected;
+        
+        //console.log(matches);
```
Comment (Elchi3) [inline test/test-doctests.js]: Remove comment

**Comment by Elchi3** [comment]
Very good work! I am happy to see this change. Some checks are really easier to read and understand now. I filed a follow-up to split tests into own files to make this even more readable.

I think another follow-up could be to introduce some utility functions as some code seems to be repeated in the tests. If there is a simple tests that uses just a `querySelectorAll` call, we could wrap it and re-use it, for example.

**Comment by Elchi3** [comment]
I addressed my comments besides the `innerHTML` call. Unsure how to replace that in an elegant way yet.

The test tab loads about:blank now, as it executes the tests way faster and then you get no warnings and errors from the real website.

**Comment by SebastianZ** [comment]
> I addressed my comments besides the `innerHTML` call. Unsure how to replace that in an elegant way yet.

Looks good to me.
At least for AMO there is no need to replace the `innerHTML` call, because the unit tests do not have to be included in the XPI. See issue #91. If you want to get rid of `innerHTML` completely, I suggest to open up another issue for it.

> The test tab loads about:blank now, as it executes the tests way faster and then you get no warnings and errors from the real website.

Very good point!
Though note, that in the future the unit tests may be extended to test them on real content. To avoid the slowness only one tab should be opened for all tests if possible. This doesn't avoid the warnings from the tab, though.

Sebastian

**Pull request merged**